### DISCUSSION TITLE
How to create a lambda layer for prisma using AWS CDK

### DISCUSSION THREAD

**Discussion by kaykhan** [prisma/prisma#22752] [Q&A]
### Question

I am using aws cdk to deploy my lambda function which requires prisma. 

I currently have an implementation of this working which simply bundles prisma with the lambda function.

i would like to make use of lambda layers to reuse the same prisma across multiple lambda functions and reduce the size of the original lambda function.

Can someone guide me through how we might do this. Unfortunately there is nothing on the docs related to aws cdk. https://www.prisma.io/docs/orm/prisma-client/deployment/serverless/deploy-to-aws-lambda

**initial**

     const MyLfn = new NodejsFunction(this, "MyLfn", {
            functionName: "ex-scheduler",
            runtime: lambda.Runtime.NODEJS_20_X,
            entry: "src/ex/scheduler/scheduler.ts",
            handler: "main",
            timeout: Duration.seconds(60),
            memorySize: 256,
            vpc: VPC,
            bundling: {
                nodeModules: ["@prisma/client", "prisma"],
                commandHooks: {
                    beforeBundling(_inputDir: string, _outputDir: string): string[] {
                        return [];
                    },
                    beforeInstall(_inputDir: string, _outputDir: string) {
                        return [];
                    },
                    afterBundling(inputDir: string, outputDir: string): string[] {
                        return [
                            `cp ${inputDir}/node_modules/.prisma/client/libquery_engine-rhel-openssl-1.0.x.so.node ${outputDir}`,
                            `cp ${inputDir}/prisma/schema.prisma ${outputDir}`,
                            "npx prisma generate",
                        ];
                    },
                },
                forceDockerBundling: false,
            },
        });.

**Comment by ludralph** [comment]
Hi @kaykhan 👋 

Thank you for raising this question.

Have you checked out this [resource](https://medium.com/ama-tech-blog/deploying-prisma-with-aws-lambda-layers-in-cdk-10608f5598c5) for deploying Prisma with AWS Lamda Layers with CDK.

**Reply by kaykhan** [reply]
I have come across this and i am attempting to reproduce but unfortunately i have not been able to, i suspect things might have changed since 2 years back. 

Do you have any updated documentation on how to solve this?

Also the way prisma is being utilised would require me to rewrite a lot of my imports.

`import { createUser } from '/opt/client';`

I am just looking to import `import { PrismaClient, Prisma } from ""`

**Reply by kaykhan** [reply]
Even that initial code i posted i come to find out is wrong, but its reused everywhere on slack and repasted by a lot of people. Kind of why i was hoping for an updated documentation on this.

Its wrong becuase that implementation regenerates @prisma and .prisma in `npx prisma generate`, this command should be done before .

**Reply by ludralph** [reply]
Hi @kaykhan 

> Do you have any updated documentation on how to solve this?

No, we do not have an updated documentation on this yet. We are tracking this using this [feature request](https://github.com/prisma/prisma/issues/12506). You may also find this [issue](https://github.com/prisma/prisma/issues/19358) of interest as well.

**Comment by josechifflet** [comment]
Hi @kaykhan. I've been struggling with this recently too.

I had an approach very similar to yours and i tryied many possible solutions but wasnt able to solve so i didnt suceed to deploy a lambda function using prisma as orm with the AWS CDK using the `NodejsFunction`

The only way i was able to achieve that was through a docker image.


```
FROM public.ecr.aws/lambda/nodejs:20 as build
WORKDIR /build
COPY package-lock.json package.json ./
COPY prisma ./prisma
RUN npm ci
COPY . .

ARG handler

# https://esbuild.github.io/api/
# also refer to: https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda-nodejs/lib/bundling.ts
RUN npx esbuild src/handlers/${handler}/handler.ts \
    --bundle \
    --minify \
    --sourcemap \
    --platform=node \
    --target=es2022 \
    --outdir=dist \
    --external:@aws-sdk/* \
    --external:@prisma/client
    
FROM public.ecr.aws/lambda/nodejs:20
WORKDIR ${LAMBDA_TASK_ROOT}

COPY package-lock.json package.json ./
COPY prisma ./prisma
RUN npm ci --omit=dev
COPY --from=build /build/dist .
COPY . .

CMD ["handler.handler"]
```

Then in your CDK:
```
new lambda.DockerImageFunction(scope, resourceId, {
      architecture: lambda.Architecture.X86_64,
      timeout: cdk.Duration.seconds(15),
      functionName: 'handler name',
      allowPublicSubnet: false,
      environment,
      memorySize: 1024,
      code: lambda.DockerImageCode.fromImageAsset("./services", { platform: ecr_assets.Platform.LINUX_AMD64 }),
    });
```

where `./services` is the path to your lambda project's and where your Dockerfile is.

Take a look to [this](https://github.com/aws-samples/prisma-lambda-cdk/blob/main/backend/Dockerfile) example. It was super helpful to me.

**Reply by kaykhan** [reply]
I appreicate this reply, but i have gotten prisma to bundle with NodejsFunction. The main point of this discussion is to understand how to get layers working.

```
   const MyFunction = new NodejsFunction(this, "MyFunction", {
            ....
            bundling: {
                externalModules: [...PowertoolsExcludeModules],
                nodeModules: [],
                commandHooks: {
                    beforeBundling(_inputDir: string, _outputDir: string): string[] {
                        return [];
                    },
                    beforeInstall(_inputDir: string, _outputDir: string) {
                        return ["npx prisma generate"];
                    },
                    afterBundling(inputDir: string, outputDir: string): string[] {
                        return [
                            `cp ${inputDir}/node_modules/.prisma/client/libquery_engine-rhel-openssl-3.0.x.so.node ${outputDir}`,
                            `cp ${inputDir}/prisma/schema.prisma ${outputDir}`,
                        ];
                    },
                },
                forceDockerBundling: false,
            },
        });


```

**Reply by josechifflet** [reply]
Awesome !

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
